FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,87 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		context.perform {
+			do {
+				if let cache = try ManagedCache.find(in: context) {
+					completion(.found(
+						feed: cache.feed
+							.compactMap { ($0 as? ManagedFeedImage) }
+							.map {
+								LocalFeedImage(id: $0.id,
+								               description: $0.imageDescription,
+								               location: $0.location,
+								               url: $0.url)
+							},
+						timestamp: cache.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		context.perform {
+			do {
+				let managedCache = try ManagedCache.getNewUniqueInstance(context: context)
+				managedCache.timestamp = timestamp
+				managedCache.feed = NSOrderedSet(array: feed.map { local in
+					let managed = ManagedFeedImage(context: context)
+					managed.id = local.id
+					managed.imageDescription = local.description
+					managed.location = local.location
+					managed.url = local.url
+					return managed
+				})
+				try context.save()
+				completion(nil)
+			} catch {
+				context.reset()
+				completion(error)
+			}
+		}
 	}
 
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		context.perform {
+			do {
+				try ManagedCache.find(in: context).map(context.delete).map(context.save)
+				completion(nil)
+			} catch {
+				context.reset()
+				completion(error)
+			}
+		}
 	}
 }
+
+public class ManagedCache: NSManagedObject {
```

Comment:
`ManagedCache` is an internal implementation detail of `CoreDataFeedStore`. It doesn't need to be public and since it is declared inside `CoreDataFeedStore.swift` it doesn't need to be `internal` as well and can be `private`. 
Also this class is not intended to be inherited thus it should be `final`
Please make the class `private final`.

```suggestion
private final class ManagedCache: NSManagedObject {
```

Reply:
Hi Danil,
Thank you for the review, it is really helpful :)
When I make ManagedCache and ManagedFeedImage private I get the error: "An NSManagedObject of class 'ManagedCache' must have a valid NSEntityDescription."
I guess this is because of functions in CoreDataHelpers.swift, they don't have access to it...?
So I will move these classes into their own files and make them internal.